File: src/focalnet/candidates.py
"""Pure NumPy crop candidate generation shared by training and inference."""

import math

import numpy as np

MAX_CANDIDATES = 128


def generate_candidates(
    image_width: int,
    image_height: int,
    aspect_ratio: float,
    *,
    positions: int = 5,
    scales: tuple[float, ...] = (1.0, 0.9, 0.8, 0.7, 0.65),
) -> np.ndarray:
    """Generate exact-ratio source-normalized crops with multiple positions and zoom levels."""
    if min(image_width, image_height, positions) <= 0:
        raise ValueError("Image dimensions and candidate positions must be positive")
    if not math.isfinite(aspect_ratio) or aspect_ratio <= 0:
        raise ValueError("Aspect ratio must be finite and positive")
    if not scales or any(not math.isfinite(scale) or not 0 < scale <= 1 for scale in scales):
        raise ValueError("Candidate scales must be finite values in (0, 1]")
    source_ratio = image_width / image_height
    if source_ratio > aspect_ratio:
        maximum_width, maximum_height = aspect_ratio / source_ratio, 1.0
    else:
        maximum_width, maximum_height = 1.0, source_ratio / aspect_ratio
    candidates = []
    for scale in scales:
        width, height = maximum_width * scale, maximum_height * scale
        xs = np.linspace(0, 1 - width, positions if width < 1 - 1e-7 else 1)
        ys = np.linspace(0, 1 - height, positions if height < 1 - 1e-7 else 1)
        for top in ys:
            for left in xs:
                candidates.append((left, top, left + width, top + height))
    result = np.unique(np.round(np.asarray(candidates, dtype=np.float32), 7), axis=0)
    if len(result) > MAX_CANDIDATES:
        raise ValueError(f"Generated {len(result)} crops; maximum is {MAX_CANDIDATES}")
    return result


def letterbox_content(box) -> np.ndarray:
    """Return normalized content bounds consumed by the crop-ranking model."""
    return (
        np.asarray(
            [box.left, box.top, box.left + box.width, box.top + box.height], dtype=np.float32
        )
        / box.size
    )

File: web/src/candidates.ts
export const MAX_CANDIDATES = 128;
export const DEFAULT_SCALES = [1.0, 0.9, 0.8, 0.7, 0.65] as const;
export const DEFAULT_POSITIONS = 5;

export type Letterbox = {
  size: number;
  left: number;
  top: number;
  width: number;
  height: number;
};

function linspace(start: number, stop: number, count: number): number[] {
  if (count === 1) {
    return [start];
  }
  const step = (stop - start) / (count - 1);
  const values = Array.from({ length: count }, (_, index) => start + step * index);
  values[count - 1] = stop;
  return values;
}

function round7(value: number): number {
  return Math.round(value * 1e7) / 1e7;
}

function uniqueSortedBoxes(boxes: number[][]): Float32Array[] {
  const seen = new Map<string, Float32Array>();
  for (const box of boxes) {
    const rounded = new Float32Array(4);
    for (let index = 0; index < 4; index += 1) {
      rounded[index] = Math.fround(round7(Math.fround(box[index])));
    }
    seen.set(rounded.join(","), rounded);
  }
  return [...seen.values()].sort((left, right) => {
    for (let index = 0; index < 4; index += 1) {
      if (left[index] < right[index]) {
        return -1;
      }
      if (left[index] > right[index]) {
        return 1;
      }
    }
    return 0;
  });
}

export function generateCandidates(
  imageWidth: number,
  imageHeight: number,
  aspectRatio: number,
  positions = DEFAULT_POSITIONS,
  scales: readonly number[] = DEFAULT_SCALES,
): Float32Array[] {
  if (Math.min(imageWidth, imageHeight, positions) <= 0) {
    throw new Error("Image dimensions and candidate positions must be positive");
  }
  if (!Number.isFinite(aspectRatio) || aspectRatio <= 0) {
    throw new Error("Aspect ratio must be finite and positive");
  }
  if (
    !scales.length ||
    scales.some((scale) => !Number.isFinite(scale) || scale <= 0 || scale > 1)
  ) {
    throw new Error("Candidate scales must be finite values in (0, 1]");
  }
  const sourceRatio = imageWidth / imageHeight;
  const maximumWidth = sourceRatio > aspectRatio ? aspectRatio / sourceRatio : 1;
  const maximumHeight = sourceRatio > aspectRatio ? 1 : sourceRatio / aspectRatio;
  const boxes: number[][] = [];
  for (const scale of scales) {
    const width = maximumWidth * scale;
    const height = maximumHeight * scale;
    const xs = linspace(0, 1 - width, width < 1 - 1e-7 ? positions : 1);
    const ys = linspace(0, 1 - height, height < 1 - 1e-7 ? positions : 1);
    for (const top of ys) {
      for (const left of xs) {
        boxes.push([left, top, left + width, top + height]);
      }
    }
  }
  const unique = uniqueSortedBoxes(boxes);
  if (unique.length > MAX_CANDIDATES) {
    throw new Error(`Generated ${unique.length} crops; maximum is ${MAX_CANDIDATES}`);
  }
  return unique;
}

export function letterboxContent(box: Letterbox): Float32Array {
  return new Float32Array([
    box.left / box.size,
    box.top / box.size,
    (box.left + box.width) / box.size,
    (box.top + box.height) / box.size,
  ]);
}

export function padBoxes(boxes: Float32Array[]): Float32Array {
  const padded = new Float32Array(MAX_CANDIDATES * 4);
  for (let index = 0; index < boxes.length; index += 1) {
    padded.set(boxes[index], index * 4);
  }
  return padded;
}
